### PR TITLE
support encoding/decoding SSB URIs semantically

### PR DESCRIPTION
**Pull request by staltz**
- [x] Code reviews
- [x] Pending on https://github.com/ssb-ngi-pointer/ssb-meta-feed-spec/pull/25 merged first
- [x] Pending on https://github.com/ssb-ngi-pointer/ssb-bfe-spec/pull/20 merged first
- [x] Pending on a new version of ssb-bfe-spec released on npm and updated here

### CODE REVIEW

File: test/00-feed.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ const { bfeNamedTypes, bfeTypes } = bfe
 tape('00 feed type', function (t) {
   const values = [
     '@6CAxOI3f+LUOVrbAl0IemqiS7ATpQvr9Mdw9LC4+Uv0=.ed25519', // classic
-    '@6CAxOI3f+LUOVrbAl0IemqiS7ATpQvr9Mdw9LC4+Uv0=.bbfeed-v1', // bendy-butt
+    'ssb:feed/bendybutt-v1/6CAxOI3f-LUOVrbAl0IemqiS7ATpQvr9Mdw9LC4-Uv0=', // bendy-butt
```
Comment (mixmix) [inline test/00-feed.js:9]: This diff makes it really nice and clear what's happening!

File: test/01-msg.js [hunk 2]
```diff
@@ -4,7 +4,7 @@ const bfe = require('../')
 tape('01 msg type', function (t) {
   const values = [
     '%HZVnEzm0NgoSVfG0Hx4gMFbMMHhFvhJsG2zK/pijYII=.sha256', // classic
-    '%HZVnEzm0NgoSVfG0Hx4gMFbMMHhFvhJsG2zK/pijYII=.bbmsg-v1', // bendy-butt
+    'ssb:message/bendybutt-v1/HZVnEzm0NgoSVfG0Hx4gMFbMMHhFvhJsG2zK_pijYII=', // bendy-butt
```
Comment (mixmix) [inline test/01-msg.js:7]: This is going to be a breaking channing in bfe.json?

Are we gonna commit to message or msg as the typeName (or allow both)?

File: package.json [hunk 3]
```diff
@@ -16,7 +16,7 @@
   },
   "dependencies": {
     "is-canonical-base64": "^1.1.1",
-    "ssb-bfe-spec": "^0.1.1",
+    "ssb-bfe-spec": "github:ssb-ngi-pointer/ssb-bfe-spec#support-ssb-uris",
```
Comment (mixmix) [inline package.json:19]: This do

**Review by mixmix — COMMENTED** [review]
The message /msg thing is my main question.

Oh, and had thought about modifying the decide to to something like decode(buf, { toURI: true}) do you can force classic things to be uris too.

I think your solution of moving sigil seems wise though

File: test/01-msg.js (hunk 2 above)
Comment (staltz) [reply test/01-msg.js:7]: SSB URIs are `ssb:message/.....` and that's a thing already deployed in Patchfox, Manyverse, Patchfoo, so I wouldn't fiddle with those. Whether to support `ssb:msg/....` opens up the question of whether we should have aliases for URIs or not. Personally, I prefer when there is one way (no need to support two things, no need to make a choice which one to use at each case, etc), and anyway I think it's a bigger discussion outside of the scope of this PR

Comment (staltz) [reply test/01-msg.js:7]: > This is going to be a breaking channing in bfe.json?

Oh, and yes

Comment (mixmix) [reply test/01-msg.js:7]: Yeah I'm fine with all that. Agree

**Comment by staltz** [comment]
@mycognosist I think this is a PR you could review, and I can drop arj (assuming he has too much to work on, but obviously if arj has comments I gladly welcome them too).

File: package.json (hunk 3 above)
Comment (mixmix) [reply package.json:19]: I think you can upgrade this to the module latest now

**Comment by mixmix** [comment]
@staltz this looks good to me, just needs module installing I think.

**Comment by staltz** [comment]
Thanks @mixmix . I won't merge this PR before updating ssb-bfe-spec to an npm version, and before marking it non-draft. Just wanted more feedback on the code, so that once ssb-bfe-spec is ready, we are ready to merge.

**Review by mycognosist — APPROVED** [review]
I can't spot anything obviously incorrect or out of place. Should be golden 👍🏻

**Comment by staltz** [comment]
Due to the tests being fundamentally changed (even if just a little), this will have be a new breaking change version.

**Pull request merged**